### PR TITLE
Before/after hooks

### LINKED ISSUE #4: Hooks
Adding some sort of `before` and `after` hooks might be useful. Especially for creating and removing fixtures or any other generated files after tests have finished.

### PR DESCRIPTION
**Pull request by vdemedes**
Fixes #4.

Hooks are completely the same as regular tests and provide the same functionality/API.
The only thing different is **when** they are executed.

### Changes

- Add `test.before()` to execute a test before all others
- Add `test.after()` to execute a test after all others
- Refactor `Runner.prototype.concurrent` and `Runner.prototype.serial` to accept array of tests, instead of getting them directly from `this.tests.concurrent` and `this.tests.serial` properties respectively
- Add tests
- Add information about before/after hooks to the readme

### Behavior

After this change, here is how Ava runs tests:

1. before hooks
2. serial tests
3. concurrent tests
4. after hooks

*Update*: If at least one before hook fails, all the following tests won't be executed.

### Usage

```js
var test = require('ava');

// shortcuts for convenience
var before = test.before;
var after = test.after;

before('before hook', function (t) {
  // this will run before everything else

  t.end();
});

after('after hook', function (t) {
  // this will run after everything else is completed

  t.end();
});

test('regular test', function (t) {
  t.end();
});
```

Output:

screen shot 2015-09-05 at 1 50 53 pm

### Tests

Tests were added to cover new `test.before()` and `test.after()` functions.

Let me know what you think!

### CODE REVIEW

File: lib/runner.js [hunk 1]
```diff
@@ -75,23 +85,23 @@ Runner.prototype.serial = function (cb) {
...
-	if (serial.length > 0 && concurrent.length > 0) {
-		this.serial(this.concurrent.bind(this, this.end.bind(this, cb)));
-		return;
-	}
-
-	this.end(cb);
+	var before = this.tests.before;
+	var after = this.tests.after;
+
+	// TODO: refactor this bullshit
```
Comment (Qix-) [inline lib/runner.js]: ;)

File: readme.md [hunk 2]
```diff
@@ -168,6 +168,35 @@ test.serial(function (t) {
 ```
 
 
+### Before/after hooks
+
+There are cases, when you need to run certain functionality to prepare or clean up after tests.
+There are `test.before()` and `test.after()` functions, which are completely the same as `test()`.
+But with their help, tests can be run before or after all tests defined using `test()`.
+Before and after hooks are executed serially.
```
Comment (Qix-) [inline readme.md]: > When setup and/or teardown is required, you can use `test.before()` and `test.after()`, used in the same manner as `test()`. The test function given to `test.before()` and `test.after()` is called before/after each test.

Comment (vdemedes) [inline readme.md]: Sometimes I forget I write docs for developers, haha

Comment (Qix-) [inline readme.md]: ;) No problem haha

Comment (vdemedes) [inline readme.md]: Will commit your version soon, thank you ;)

**Comment by Qix-** [comment]
Other than the two things I saw (I'm sure @sindresorhus will see something else), it looks good to me. Seems like a clean implementation.

File: readme.md [hunk 3]
```diff
@@ -168,6 +168,35 @@ test.serial(function (t) {
...
+
+```js
+// shortcuts for convenience
+var before = test.before;
+var after = test.after;
+
+before (function (t) {
```
Comment (arthurvr) [inline readme.md]: `before(function (t) {` (spacing)

**Comment by vdemedes** [comment]
Yeah, looks good to me too, except `Runner.prototype.run`.

File: readme.md (hunk 3 above)
Comment (vdemedes) [inline readme.md]: thanks, will change ;)

File: lib/runner.js (hunk 1 above)
Comment (sindresorhus) [inline lib/runner.js]: Relevant: https://github.com/sindresorhus/ava/issues/38

// @vdemedes

Comment (vdemedes) [inline lib/runner.js]: Yeah, I thought about converting `this.serial` and `this.concurrent` to return promises, but decided postpone that. Wanted to hear feedback on a PR and thought that you might not like too much changes at once, haha

Comment (sindresorhus) [inline lib/runner.js]: Yeah, promisification should be a separate PR.

Comment (vdemedes) [inline lib/runner.js]: I could do that after this PR gets merged, no probs

**Comment by sindresorhus** [comment]
Looks superb to me. Can you fix the merge conflict? :D

**Comment by vdemedes** [comment]
@sindresorhus Yeah, sure, already on it ;)

**Comment by sindresorhus** [comment]
Excellent! Thanks again @vdemedes for this awesome contribution :)

687474703a2f2f74636c686f73742e636f6d2f36763641355a392e676966

**Comment by vdemedes** [comment]
@sindresorhus You're welcome! I am really enjoying contributing to `ava` and `got`, amazing stuff. Switched almost all my shit to them from `request` and `mocha`.

P.S. That gif is amazing :D

**Comment by Qix-** [comment]
Keep them coming @vdemedes :)
